src/ngx_http_lua_socket_tcp.c

```diff
@@ -2963,6 +2981,225 @@ ngx_http_lua_socket_tcp_send(lua_State *L)
     return lua_yield(L, 0);
 }
 
+#ifndef NGX_LUA_NO_FFI_API
+int
+ngx_http_lua_socket_tcp_ffi_set_buf(ngx_http_request_t *r,
+    ngx_http_lua_socket_tcp_upstream_t *u, void *cdata_ctx, size_t len,
+    char **err)
+{
+    ngx_chain_t                         *cl;
+    ngx_http_lua_ctx_t                  *ctx;
+    ngx_buf_t                           *b;
+    ngx_http_lua_loc_conf_t             *llcf;
+
+    dd("tcp set buf: u=%p, u->write_closed=%d", u, (unsigned) u->write_closed);
+    if (u == NULL || u->peer.connection == NULL || u->write_closed) {
+        llcf = ngx_http_get_module_loc_conf(r, ngx_http_lua_module);
+
+        if (llcf->log_socket_errors) {
+            ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+                          "attempt to send data on a closed socket: u:%p, "
+                          "c:%p, ft:%d eof:%d",
+                          u, u ? u->peer.connection : NULL,
+                          u ? (int) u->ft_type : 0, u ? (int) u->eof : 0);
+        }
+
+        *err = "closed";
+        return NGX_ERROR;
+    }
+
+    if (u->request != r) {
+        *err = "bad request";
+        return NGX_ERROR;
+    }
+
+    if (u->conn_waiting) {
+        *err = "socket busy connecting";
+        return NGX_ERROR;
+    }
+
+    if (u->write_waiting) {
+        *err = "socket busy writing";
+        return NGX_ERROR;
+    }
+
+    if (u->raw_downstream
+        && (r->connection->buffered & NGX_HTTP_LOWLEVEL_BUFFERED)) {
```

Comment:
When the if condition is in multiple lines, the `{` should be on a separate line by itself. See the guide mentioned above.